packages/transformers/webext/src/schema.js

```diff
@@ -0,0 +1,402 @@
+// @flow strict-local
+import type {SchemaEntity} from '@parcel/utils';
+import {extname} from 'path';
+
+// Non-comprehensive, but should cover the vast majority of users
+// SVGs not supported in WebKit/Blink
+const IMG_EXT = new Set(['jpg', 'jpeg', 'png', 'svg', 'webp', 'gif']);
+
+const validateVersion = (ver: string): ?string => {
+  const parts = ver.split('.', 5);
+  if (parts.length == 5) return 'Extension versions to have at most three dots';
+  // greater than to allow for NaN to fail
+  if (parts.every(part => +part[0] > 0 && +part < 65536)) return;
+  return 'Extension versions must be dot-separated integers between 0 and 65535';
+};
+
+const validateImage = (name: string): ?string => {
+  const ext = extname(name).slice(1);
+  if (!IMG_EXT.has(ext))
+    return `Extensions do not support ${ext} images and icons`;
+};
+
+const icons: SchemaEntity = {
+  type: 'object',
+  properties: {},
+  additionalProperties: {
+    type: 'string',
+    __validate: validateImage,
+  },
+};
+
+const actionProps = {
+  // FF only
+  browser_style: {type: 'boolean'},
+  // You can also have a raw string, but not in Edge, apparently...
+  default_icon: icons,
```

Comment:
According to documentation, [`browser_action.icons` should be `string` or `object`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/browser_action):

I guess something like:

```ts
const icons: SchemaOneOf = {
  oneOf: [{
    type: 'object',
    properties: {},
    additionalProperties: {
      type: 'string',
      __validate: validateImage,
    },
  },
  {
    type: 'string',
    __validate: validateImage,
  }]
};
```

Reply:
I'm aware, this is intentional to support Edge. The goal was maximizing compat.

Reply:
Edge (which is the same API as chrome) [supports both `string` or `object`](https://developer.chrome.com/extensions/browserAction) as well. To maximize compatibility we should support both cases too.

Reply:
[This](https://docs.microsoft.com/en-us/microsoft-edge/extensions/api-support/supported-manifest-keys#browser_action-or-page_action-keys) disagrees.

Reply:
This if for the "legacy" Edge up to version 44: https://docs.microsoft.com/en-us/microsoft-edge/extensions/api-support/supported-manifest-keys#browser_action-or-page_action-keys

This applies to the modern Edge based on Chromium: https://developer.chrome.com/extensions/browserAction

Reply:
Good catch, very interesting because the extension I'm testing against uses a string:

https://github.com/wachunei/directUC/blob/9583cbc832f802977437cfbc04aa0441b17d2bf9/src/manifest/base.json#L22

And it works (passes Edge store validations), it is [published here.](https://microsoftedge.microsoft.com/addons/detail/directuc/hnlkkfgdllbagiaommmbnagdbnciddfl).

From the set of restrictions though, should we just use the less restrictive? Since this is not a tool for manifest validations, in case Chrome/Edge enforces the array syntax should be up to the dev to update it. What do you think?

Reply:
> this is not a tool for manifest validations, ... should it be up to the dev to update it

👍

Reply:
That's true. In hindsight, the massive schema may not have been necessary per se since validation isn't the goal, but I suppose it (mostly) doesn't hurt.

Reply:
I think it is essential actually, super helpful for the user to catch errors early in development 👌